### PR TITLE
Issue/8321 add api endpoint for scheduler self check

### PR DESCRIPTION
**Pull request by Hugo-Inmanta**
# Description

closes https://github.com/inmanta/inmanta-core/issues/8321

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )
- [ ] If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)

### CODE REVIEW

File: src/inmanta/agent/agent_new.py [hunk 1]
```diff
@@ -28,17 +28,17 @@
...
+from inmanta.protocol import SessionEndpoint, TypedClient, methods, methods_v2
 from inmanta.server.services.databaseservice import DatabaseMonitor
 from inmanta.types import Apireturn
 from inmanta.util import ensure_directory_exist, join_threadpools
 
 LOGGER = logging.getLogger("inmanta.scheduler")
 
 
-class Agent(SessionEndpoint):
+class Agent(SessionEndpoint, TypedClient):
```
Comment (wouterdb) [inline src/inmanta/agent/agent_new.py:41]: 1. can you do this? 
2. What does this mean?

File: src/inmanta/data/model.py [hunk 2]
```diff
@@ -898,6 +898,19 @@ def has_source(self) -> bool:
...
+    """
+    Status report for the scheduler self-check
+
+    :param resource_state: In-memory representation of the resources in the scheduler
+    :param discrepancies: Discrepancies between the in-memory representation of the resources
+        and their state in the database.
+    """
+
+    resource_state: dict[ResourceIdStr, typing.Any]
+    discrepancies: dict[ResourceIdStr, typing.Any]
```
Comment (wouterdb) [inline src/inmanta/data/model.py:911]: It would be nicer if the second arguments are typed, not crucial

File: src/inmanta/deploy/scheduler.py [hunk 3]
```diff
@@ -372,6 +382,107 @@ async def get_facts(self, resource: dict[str, object]) -> None:
...
+        """
+        Check that the state of the resources in the DB corresponds
+        to the internal state of the scheduler and return a SchedulerStatusReport
+        object containing the internal state and any discrepancies between this
+        state and that of the DB.
+        """
+
+        resources_in_db: Mapping[ResourceIdStr, ResourceDetails]
+
+        try:
```
Comment (wouterdb) [inline src/inmanta/deploy/scheduler.py:395]: We may want to grab some lock to make sure we read one version of the data?

File: src/inmanta/deploy/scheduler.py [hunk 4]
```diff
@@ -372,6 +382,107 @@ async def get_facts(self, resource: dict[str, object]) -> None:
...
+            # Resources only present in the DB but missing from the scheduler
+            only_in_db = resources_in_db.keys() - self._state.resource_state.keys()
+            for rid in only_in_db:
+                discrepancy_map[rid] = [
+                    Discrepancy(
+                        expected=(
+                            f"Resource {rid} is present in the DB (model version {latest_version}). "
+                            "It is expected in the scheduler's resource_state map."
+                        ),
+                        actual=f"Resource {rid} is missing from the scheduler's resource_state map.",
```
Comment (wouterdb) [inline src/inmanta/deploy/scheduler.py:437]: I wonder if we can split the context into different fields in the `Discrepancy` i.e. such that is has a field for 'rid', an optional for 'field' (which field is it about) and then expected actual

File: src/inmanta/deploy/scheduler.py [hunk 5]
```diff
@@ -372,6 +382,107 @@ async def get_facts(self, resource: dict[str, object]) -> None:
...
+                            Discrepancy(expected=db_compliance_status, actual=scheduler_resource_state.status)
+                        )
+
+                if resource_discrepancies:
+                    discrepancy_map[rid] = resource_discrepancies
+
+            return discrepancy_map
+
+        discrepancy_map = _build_discrepancy_map(resources_in_db)
+        return SchedulerStatusReport(resource_state=self._state.resource_state, discrepancies=discrepancy_map)
```
Comment (wouterdb) [inline src/inmanta/deploy/scheduler.py:484]: We may want to pack the DB state in here as well.

File: src/inmanta/server/services/orchestrationservice.py [hunk 6]
```diff
@@ -1305,6 +1306,20 @@ async def get_diff_of_versions(
...
+        env: data.Environment,
+    ) -> SchedulerStatusReport:
+        await self.autostarted_agent_manager._ensure_scheduler(env.id)
+        client = self.agentmanager_service.get_agent_client(env.id, const.AGENT_SCHEDULER_ID)
+        assert client is not None
+        status = await client.trigger_get_status(env.id)
+        assert status.code == 200
+        assert status.result is not None
+        resp = SchedulerStatusReport.model_validate(status.result["data"])
+        return resp
```
Comment (wouterdb) [inline src/inmanta/server/services/orchestrationservice.py:1321]: There assert should not be asserts, but raise proper errors with a proper http code

**Review by wouterdb — COMMENTED** [review]
Looks good overall

File: src/inmanta/agent/agent_new.py (hunk 1 above)
Comment (Hugo-Inmanta) [reply src/inmanta/agent/agent_new.py]: Removed, it was an artifact from experimentation with `TypedClient`

File: src/inmanta/deploy/scheduler.py (hunk 5 above)
Comment (Hugo-Inmanta) [reply src/inmanta/deploy/scheduler.py]: How much of it should we pack ? I plopped everything from the `ResourceDetails` in there but this is maybe overkill ?

File: src/inmanta/data/model.py (hunk 2 above)
Comment (Hugo-Inmanta) [reply src/inmanta/data/model.py]: I don't get how to type these, there's probably something obvious I'm missing. 
e.g. for `resource_state` 
Should we import ResourceState from inmanta.deploy.state and type it as such here ? I tried but I was running into circular imports but I can have another look if that's the way to go.

`    resource_state: dict[ResourceIdStr, ResourceState]`

File: src/inmanta/deploy/scheduler.py [hunk 7]
```diff
@@ -372,6 +372,139 @@ async def get_facts(self, resource: dict[str, object]) -> None:
...
+            except KeyError:
+                return SchedulerStatusReport(scheduler_state={}, db_state={}, discrepancies={})
+
+            if latest_version != self._state.version:
+                LOGGER.info(
+                    "Cannot compare scheduler status and database status because of "
+                    "model version mismatch (%s and %s respectively). "
+                    "The scheduler might not have processed this new version yet.",
+                    self._state.version,
+                    latest_version,
```
Comment (wouterdb) [inline src/inmanta/deploy/scheduler.py:499]: This message should go in the report as well.

File: src/inmanta/deploy/scheduler.py [hunk 8]
```diff
@@ -372,6 +372,139 @@ async def get_facts(self, resource: dict[str, object]) -> None:
...
+        """
+        Check that the state of the resources in the DB corresponds
+        to the internal state of the scheduler and return a SchedulerStatusReport
+        object containing the internal state and any discrepancies between this
+        state and that of the DB.
+        """
+
+        resources_in_db: Mapping[ResourceIdStr, ResourceDetails]
+
+        def _build_discrepancy_map(
```
Comment (wouterdb) [inline src/inmanta/deploy/scheduler.py:385]: This method may take a long time. I wonder if we should yield to the ioloop ('await sleep(0)`) every now and them to keep the session alive ? (e.g. every 1000 loop iterations) or run this on a thread

File: src/inmanta/data/model.py [hunk 9]
```diff
@@ -898,6 +898,43 @@ def has_source(self) -> bool:
...
+    Status report for the scheduler self-check
+
+    :param scheduler_state: In-memory representation of the resources in the scheduler
+    :param db_state: Desired state of the resources as persisted in the database
+    :param discrepancies: Discrepancies between the in-memory representation of the resources
+        and their state in the database.
+    """
+
+    scheduler_state: Mapping[ResourceIdStr, typing.Any]
+    db_state: Mapping[ResourceIdStr, typing.Any]
```
Comment (wouterdb) [inline src/inmanta/data/model.py:934]: Would it be difficult to have these typed as well? i.e. if easy, I would, if the existing types don't work, I would document why it is any.

**Review by wouterdb — APPROVED** [review]
Cool!

File: src/inmanta/data/model.py [hunk 10]
```diff
@@ -898,6 +898,44 @@ def has_source(self) -> bool:
...
 
+class Discrepancy(BaseModel):
+    """
+    Records a discrepancy between the state as persisted in the database and
+    the in-memory state in the scheduler.
+
+    :param rid: If set, this discrepancy is specific to this resource.
+    :param field: If set, specifies on which field this discrepancy was detected.
+        If left unset, and a rid is specified, the discrepancy was detected on the
+        resource level i.e. it is missing from either the db or the scheduler.
```
Comment (jptrindade) [inline src/inmanta/data/model.py:909]: You could also mention the case of `model_version`

File: src/inmanta/protocol/methods_v2.py [hunk 11]
```diff
@@ -392,6 +392,39 @@ def get_api_docs(format: Optional[ApiDocsFormat] = ApiDocsFormat.swagger) -> Ret
...
+    arg_options=methods.ENV_OPTS,
+    client_types=[],
+    api_version=2,
+)
+def trigger_get_status(tid: uuid.UUID) -> model.SchedulerStatusReport:
+    """
+    Get a snapshot of the scheduler state
+
+    :param tid: The id of the environment.
+    """
```
Comment (jptrindade) [inline src/inmanta/protocol/methods_v2.py:425]: I don't quite get the difference between these two endpoints

**Review by jptrindade — CHANGES_REQUESTED** [review]
Is it possible (maybe even in another PR) to make this call at the end of every test case and assert that `discrepancies` is empty? Making it only run in the nightly if it is too heavy

File: src/inmanta/protocol/methods_v2.py (hunk 11 above)
Comment (Hugo-Inmanta) [reply src/inmanta/protocol/methods_v2.py:425]: `get_scheduler_status` is the outer one: api on the server to request the status, (the orchestration service responds to this one) and it in turns calls the inner `trigger_get_status` to retrieve the status (the new agent responds to that one)

File: src/inmanta/data/model.py (hunk 10 above)
Comment (Hugo-Inmanta) [reply src/inmanta/data/model.py:909]: Updated docstring.

File: src/inmanta/data/model.py (hunk 9 above)
Comment (Hugo-Inmanta) [reply src/inmanta/data/model.py]: As discussed, typed as object + added a comment

**Review by jptrindade — APPROVED** [review]
